File: sdk/typescript/tests/exceptions.test.ts
/**
 * Tests for exception classes
 */

import { A2AError, AuthenticationError, ValidationError, NotFoundError } from '../src/exceptions';

describe('Exceptions', () => {
  describe('A2AError', () => {
    it('should create A2AError', () => {
      const error = new A2AError('Test error');
      expect(error).toBeInstanceOf(Error);
      expect(error).toBeInstanceOf(A2AError);
      expect(error.message).toBe('Test error');
      expect(error.name).toBe('A2AError');
    });

    it('should create A2AError with details', () => {
      const details = { code: 'ERR001', field: 'name' };
      const error = new A2AError('Test error', details);
      expect(error.details).toEqual(details);
    });
  });

  describe('AuthenticationError', () => {
    it('should create AuthenticationError', () => {
      const error = new AuthenticationError('Authentication failed');
      expect(error).toBeInstanceOf(Error);
      expect(error).toBeInstanceOf(A2AError);
      expect(error).toBeInstanceOf(AuthenticationError);
      expect(error.message).toBe('Authentication failed');
      expect(error.name).toBe('AuthenticationError');
    });

    it('should create AuthenticationError with details', () => {
      const details = { code: 'AUTH_001' };
      const error = new AuthenticationError('Auth failed', details);
      expect(error.details).toEqual(details);
    });
  });

  describe('ValidationError', () => {
    it('should create ValidationError', () => {
      const error = new ValidationError('Validation failed');
      expect(error).toBeInstanceOf(Error);
      expect(error).toBeInstanceOf(A2AError);
      expect(error).toBeInstanceOf(ValidationError);
      expect(error.message).toBe('Validation failed');
      expect(error.name).toBe('ValidationError');
    });

    it('should create ValidationError with details', () => {
      const details = { field: 'name', reason: 'required' };
      const error = new ValidationError('Invalid field', details);
      expect(error.details).toEqual(details);
    });
  });

  describe('NotFoundError', () => {
    it('should create NotFoundError', () => {
      const error = new NotFoundError('Resource not found');
      expect(error).toBeInstanceOf(Error);
      expect(error).toBeInstanceOf(A2AError);
      expect(error).toBeInstanceOf(NotFoundError);
      expect(error.message).toBe('Resource not found');
      expect(error.name).toBe('NotFoundError');
    });

    it('should create NotFoundError with details', () => {
      const details = { resource: 'agent', id: 'agent-123' };
      const error = new NotFoundError('Agent not found', details);
      expect(error.details).toEqual(details);
    });
  });

  describe('Error inheritance', () => {
    it('should properly inherit from Error', () => {
      const errors = [
        new A2AError('Base error'),
        new AuthenticationError('Auth error'),
        new ValidationError('Validation error'),
        new NotFoundError('Not found error'),
      ];

      errors.forEach((error) => {
        expect(error).toBeInstanceOf(Error);
        expect(error.stack).toBeDefined();
      });
    });

    it('should properly inherit from A2AError', () => {
      const authError = new AuthenticationError('Auth error');
      const validationError = new ValidationError('Validation error');
      const notFoundError = new NotFoundError('Not found error');

      expect(authError).toBeInstanceOf(A2AError);
      expect(validationError).toBeInstanceOf(A2AError);
      expect(notFoundError).toBeInstanceOf(A2AError);
    });
  });
});

File: sdk/go/pkg/a2areg/exceptions_test.go
package a2areg

import (
	"errors"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestA2AError(t *testing.T) {
	err := NewA2AError("Test error", map[string]interface{}{"code": "ERR001"})
	assert.Error(t, err)
	assert.Equal(t, "Test error", err.Error())
	assert.Equal(t, "ERR001", err.Details["code"])
}

func TestA2AError_Unwrap(t *testing.T) {
	baseErr := errors.New("base error")
	a2aErr := &A2AError{
		Message: "Test error",
		Err:     baseErr,
	}

	assert.Equal(t, baseErr, a2aErr.Unwrap())
	assert.Contains(t, a2aErr.Error(), "base error")
}

func TestAuthenticationError(t *testing.T) {
	err := NewAuthenticationError("Authentication failed", nil)
	assert.Error(t, err)
	assert.IsType(t, &AuthenticationError{}, err)
	assert.NotNil(t, err.A2AError)
	assert.Equal(t, "Authentication failed", err.Error())
}

func TestValidationError(t *testing.T) {
	err := NewValidationError("Validation failed", map[string]interface{}{"field": "name"})
	assert.Error(t, err)
	assert.IsType(t, &ValidationError{}, err)
	assert.NotNil(t, err.A2AError)
	assert.Equal(t, "Validation failed", err.Error())
	assert.Equal(t, "name", err.Details["field"])
}

func TestNotFoundError(t *testing.T) {
	err := NewNotFoundError("Resource not found", map[string]interface{}{"resource": "agent"})
	assert.Error(t, err)
	assert.IsType(t, &NotFoundError{}, err)
	assert.NotNil(t, err.A2AError)
	assert.Equal(t, "Resource not found", err.Error())
}

func TestRateLimitError(t *testing.T) {
	err := NewRateLimitError("Rate limit exceeded", map[string]interface{}{"retry_after": 60})
	assert.Error(t, err)
	assert.IsType(t, &RateLimitError{}, err)
}

func TestServerError(t *testing.T) {
	err := NewServerError("Server error", nil)
	assert.Error(t, err)
	assert.IsType(t, &ServerError{}, err)
}

func TestErrorInheritance(t *testing.T) {
	tests := []struct {
		name string
		err  error
	}{
		{"AuthenticationError", NewAuthenticationError("Auth error", nil)},
		{"ValidationError", NewValidationError("Validation error", nil)},
		{"NotFoundError", NewNotFoundError("Not found error", nil)},
		{"RateLimitError", NewRateLimitError("Rate limit error", nil)},
		{"ServerError", NewServerError("Server error", nil)},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Direct check that it's an A2AError
			switch e := tt.err.(type) {
			case *AuthenticationError:
				assert.NotNil(t, e.A2AError)
			case *ValidationError:
				assert.NotNil(t, e.A2AError)
			case *NotFoundError:
				assert.NotNil(t, e.A2AError)
			case *RateLimitError:
				assert.NotNil(t, e.A2AError)
			case *ServerError:
				assert.NotNil(t, e.A2AError)
			default:
				t.Fatalf("%s should be one of the error types", tt.name)
			}
		})
	}
}

